hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/ReconRestServletModule.java

```diff
@@ -32,73 +34,111 @@
-      this.paths = paths;
-    }
+  private void configureApi(String baseApiPath, String... packages) {
+    StringBuilder sb = new StringBuilder();
+    Set<String> adminEndpoints = new HashSet<>();
 
-    private void checkIfPackageExistsAndLog(String pkg) {
-      String resourcePath = pkg.replace(".", "/");
-      URL resource = getClass().getClassLoader().getResource(resourcePath);
-      if (resource != null) {
-        LOG.info("rest({}).packages({})", paths, pkg);
-      } else {
-        LOG.info("No Beans in '{}' found. Requests {} will fail.", pkg, paths);
+    for (String pkg : packages) {
+      if (sb.length() > 0) {
+        sb.append(',');
+      }
+      checkIfPackageExistsAndLog(pkg, baseApiPath);
+      sb.append(pkg);
+      // Check for classes marked as admin only that will need an extra
+      // filter applied to their path.
+      Reflections reflections = new Reflections(pkg,
+          new TypeAnnotationsScanner(), new SubTypesScanner());
+      Set<Class<?>> adminEndpointClasses =
+          reflections.getTypesAnnotatedWith(AdminOnly.class);
+      adminEndpointClasses.stream()
+          .map(clss -> clss.getAnnotation(Path.class).value())
+          .forEachOrdered(adminEndpoints::add);
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Registered the following endpoint classes as admin only: {}",
+            adminEndpointClasses);
       }
     }
+    Map<String, String> params = new HashMap<>();
+    params.put("javax.ws.rs.Application",
+        GuiceResourceConfig.class.getCanonicalName());
+    if (sb.length() > 0) {
+      params.put("jersey.config.server.provider.packages", sb.toString());
+    }
+    bind(ServletContainer.class).in(Scopes.SINGLETON);
 
-    @Override
-    public void packages(String... packages) {
-      StringBuilder sb = new StringBuilder();
+    serve(baseApiPath + "/*").with(ServletContainer.class, params);
+    addFilters(baseApiPath, adminEndpoints);
+  }
 
-      for (String pkg : packages) {
-        if (sb.length() > 0) {
-          sb.append(',');
-        }
-        checkIfPackageExistsAndLog(pkg);
-        sb.append(pkg);
+  private void addFilters(String basePath, Set<String> subPaths) {
+    if (OzoneSecurityUtil.isHttpSecurityEnabled(conf)) {
+      filter(basePath + "/*").through(ReconAuthFilter.class);
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Added authentication filter to all paths under {}",
+            basePath);
       }
-      Map<String, String> params = new HashMap<>();
-      params.put("javax.ws.rs.Application",
-          GuiceResourceConfig.class.getCanonicalName());
-      if (sb.length() > 0) {
-        params.put("jersey.config.server.provider.packages", sb.toString());
+
+      boolean aclEnabled = conf.getBoolean(OZONE_ACL_ENABLED,
+          OZONE_ACL_ENABLED_DEFAULT);
+      if (aclEnabled) {
+        for (String path: subPaths) {
+          filter(basePath + path).through(ReconAdminFilter.class);
```

Comment:
Ideally, instead of string concat it is safer to use UrlBuilder which will take care of appending separators.

Reply:
I figured a utility existed somewhere for this I just couldn't find it. Switched to using UriBuilder to build the paths.